test/unit/org/apache/cassandra/io/sstable/SSTableReaderTest.java

```diff
@@ -400,6 +403,51 @@ public void testOnDiskSizeForRanges()
         }
     }
 
+    @Test
+    public void testAssertionsOnDiskSizeForPartitionPositions()
+    {
+        Keyspace keyspace = Keyspace.open(KEYSPACE1);
+        ColumnFamilyStore cfs = keyspace.getColumnFamilyStore(CF_COMPRESSED);
+        SSTableReader sstable = getNewSSTable(cfs);
+        partitioner = sstable.getPartitioner();
+
+        assertThatThrownBy(() -> sstable.onDiskSizeForPartitionPositions(Collections.singleton(new PartitionPositionBounds(0, 0))))
+                .isInstanceOf(AssertionError.class);
+        assertThatThrownBy(() -> sstable.onDiskSizeForPartitionPositions(Collections.singleton(new PartitionPositionBounds(-1, 0))))
+                .isInstanceOf(AssertionError.class);
+        assertThatThrownBy(() -> sstable.onDiskSizeForPartitionPositions(Collections.singleton(new PartitionPositionBounds(2, 1))))
+                .isInstanceOf(AssertionError.class);
+    }
+
+    @Test
+    public void testDoNotFailOnChunkEndingPosition()
+    {
+        Keyspace keyspace = Keyspace.open(KEYSPACE1);
+        ColumnFamilyStore cfs = keyspace.getColumnFamilyStore(CF_COMPRESSED);
+        int count = 530;
+        // insert data and compact to a single sstable
+        for (int j = 0; j < count; j++)
+        {
+            new RowUpdateBuilder(cfs.metadata(), 15000, k0(j))
+                    .clustering("0")
+                    .add("val", ByteBufferUtil.EMPTY_BYTE_BUFFER)
+                    .build()
+                    .applyUnsafe();
+        }
+        cfs.forceBlockingFlush(UNIT_TESTS);
+        cfs.forceMajorCompaction();
+        SSTableReader sstable = cfs.getLiveSSTables().iterator().next();
+
+        partitioner = sstable.getPartitioner();
+        int chunkLength = sstable.getCompressionMetadata().parameters.chunkLength();
+        long chunkCount = (long) Math.ceil((double) sstable.uncompressedLength() / chunkLength);
+        assertEquals(2, chunkCount);
+        long endPosition = chunkLength * chunkCount;
+
+        long size = sstable.onDiskSizeForPartitionPositions(Collections.singleton(new PartitionPositionBounds(0, endPosition)));
+        assertNotEquals(0, size);
```

Comment:
I would add `if (sstable.uncompressedLength() > endPosition) assertTrue(size < sstable.compressedLength());` to check that this span does not include the last chunk.

If you specifically want to check the out-of-bounds case, you can try making an sstable with `chunkLength` many rows. AFAIR from `testOnDiskSizeForRanges`, the row size in these tests did end up fixed, thus this should get us an aligned length.

Reply:
Good advice, I have reworked the test. Now it reads the disk size by passing a range rather than a position and without a fix it fails on the end position. Creating `chunkLength` rows did the trick.